__OPTIMIZATION FOR LARGE SCALE DATA: Second Assignment (NLP)__ - Francesca Sallicati 100341618

Consider the data from the Spanish electrical system used in class or any other database of your
interest including one dependent variable $Y$ and at least 5 independent variables $X$, with a minimum
of 500 observations. We seek to adjust a multiple linear regression model to explain variable
Y as a function of the other variables $X$, i.e., $(Y = β'X+\epsilon)$ by using “Ridge Regression”:

$$\min(|| y - Xβ ||^2_{2}+ρ || β ||^2_{2}) $$ for $β$, where ρ is a parameter.

__a) (2 points) Estimate the value of the regression coefficients by implementing the analytical
solution.__



__SOLUTION__<br>
I chose to work on the Spanish eletrical system data and to use a value for the parameters of the penalty term equal to 1.<br>

By derivating the function in terms of $\beta$ we obtain:

$$\frac{\delta}{\delta \beta}((Y - X\beta)'(Y-X\beta) + \rho \beta'\beta))=-2X'(Y-\beta' X) + 2\rho\beta = 0 $$ 
$$ X'Y = X'X\beta + \rho \beta$$ 
$$\beta = (X'X+ \rho I )^{-1}X' Y$$

whose minimum is given by: $$\beta = (X'X+ \rho I )^{-1}X' Y$$


In [1]:
%matplotlib inline
import numpy as np
import time
import pandas as pd
import matplotlib.pyplot as plt

price=pd.read_csv('price.csv',delimiter=';',usecols=['value'])
coal=pd.read_csv('coal.csv',delimiter=';',usecols=['value'])
gas=pd.read_csv('gas.csv',delimiter=';',usecols=['value'])
wind=pd.read_csv('wind.csv',delimiter=';',usecols=['value'])
hidro=pd.read_csv('hidro.csv',delimiter=';',usecols=['value'])
nuclear=pd.read_csv('nuclear.csv',delimiter=';',usecols=['value'])
solar=pd.read_csv('solar.csv',delimiter=';',usecols=['value'])
cogen=pd.read_csv('cogen.csv',delimiter=';',usecols=['value'])
demand=pd.read_csv('demand.csv',delimiter=';',usecols=['value'])
df=pd.concat([price, coal, gas, wind, hidro, nuclear, solar, cogen, demand],axis=1)
df.columns=['price', 'coal', 'gas', 'wind', 'hidro', 'nuclear', 'solar', 'cogen', 'demand']

normalized_df=(df-df.mean())/df.std()
normalized_df

price      coal       gas      wind     hidro   nuclear     solar  \
0    -0.843676 -0.863129 -1.032348  0.429243 -0.291410  1.504478 -0.462052   
1    -1.140834 -1.183432 -1.145887  0.464282 -0.526312  1.510829 -0.488742   
2    -1.486618 -1.435346 -1.258767  0.469869 -0.952336  1.515959 -0.546745   
3    -1.609803 -1.498452 -1.278001  0.372891 -0.936070  1.516936 -0.588177   
4    -1.655187 -1.509630 -1.271681  0.293776 -0.890256  1.523287 -0.684336   
5    -1.609803 -1.420612 -1.206778  0.146101 -0.859893  1.524752 -0.754094   
6    -1.140834 -1.262187 -1.099614  0.024636 -0.297103  1.520355 -0.769414   
7    -0.580016 -1.034357 -0.908422 -0.094622  0.106962  1.523775 -0.765367   
8    -0.359579 -1.029276 -0.741961 -0.189876  0.500455  1.524997 -0.630379   
9    -0.303389 -0.854288 -0.556649 -0.397835  1.296386  1.525729 -0.103624   
10   -0.386593 -1.059762 -0.579126 -0.458464  1.496859  1.524264  0.912790   
11   -0.360660 -1.124392 -0.603142 -0.354933  1.124106  1.516447  1.636390   
12   -0.277455 -1.086284 -0.601493 -0.318790  1.216549  1.520355  1.937489   
13   -0.225588 -1.022569 -0.572147 -0.272922  1.181171  1.513272  2.042127   
14   -0.386593 -1.074192 -0.636665 -0.083310  0.594660  1.513760  2.096181   
15   -0.603789 -1.328849 -0.756744  0.262324  0.100321  1.512783  2.006381   
16   -0.827468 -1.333219 -0.859456  0.608302 -0.014759  1.507165  1.831888   
17   -1.016568 -1.381996 -0.901883  0.976076 -0.148543  1.508142  1.573473   
18   -1.140834 -1.502009 -1.036415  1.378614 -0.308353  1.510096  1.200881   
19   -1.133270 -1.526804 -0.992615  1.684932 -0.379515  1.511806  0.601381   
20   -0.654576 -1.474775 -0.917050  1.775082  0.064943  1.517424 -0.131759   
21   -0.602708 -1.318078 -0.817085  1.865646  0.300117  1.518157 -0.408770   
22   -1.286712 -1.530360 -1.084941  1.818674  0.113604  1.515226 -0.444131   
23   -1.769728 -1.700166 -1.245192  1.621130 -0.070197  1.503012 -0.451068   
24   -1.790259 -1.887247 -1.240411  1.533670 -0.457997  1.505699 -0.480455   
25   -1.891833 -2.038456 -1.191171  1.513598 -0.865044  1.512539 -0.530655   
26   -1.978279 -2.015897 -1.201557  1.079608 -1.362636  1.515470 -0.581336   
27   -2.102545 -2.027176 -1.147865  0.897721 -1.608789  1.516203 -0.642808   
28   -2.136043 -2.069450 -1.263987  0.953246 -1.686051  1.516691 -0.670846   
29   -2.123076 -2.070466 -1.192545  0.871234 -1.663144  1.521088 -0.753998   
...        ...       ...       ...       ...       ...       ...       ...   
2155  2.287828  1.609469  2.112013  0.781084  2.120239 -1.129491 -0.780784   
2156  2.322406  1.663429  2.231597  0.881857  1.789505 -1.128025 -0.814218   
2157  2.345098  1.704991  1.941979  0.938416  1.700180 -1.127536 -0.815181   
2158  1.509814  1.733445  1.384945  1.073676  0.137460 -1.122407 -0.815470   
2159  0.978172  1.706414  0.922931  1.006425 -0.371111 -1.121185 -0.815470   
2160  0.794474  1.639244  0.297037  0.947107 -0.441053 -1.122162 -0.815470   
2161  0.365486  1.645849 -0.507463  0.771980 -0.650744 -1.121185 -0.815470   
2162  0.068328  1.702857 -0.692445  0.700384 -0.684224 -1.124361 -0.815470   
2163 -0.125094  1.601645 -0.721297  0.702246 -1.563381 -1.122162 -0.815470   
2164 -0.299067  1.522382 -0.731574  0.660999 -2.190963 -1.114590 -0.815663   
2165 -0.319598  1.539454 -0.594349  0.560020 -2.292487 -1.114590 -0.815663   
2166 -0.092677  1.648898 -0.476908  0.522428 -1.147388 -1.119720 -0.815470   
2167  0.954399  1.681212 -0.120573  0.558295 -0.417875 -1.122407 -0.815470   
2168  2.131144  1.737814  1.260910  0.583057  0.042306 -1.120697 -0.815567   
2169  2.305117  1.776531  1.979734  0.497735  0.585307 -1.126315 -0.764789   
2170  2.395885  1.785677  1.992099  0.376063  1.251926 -1.134864 -0.253066   
2171  2.395885  1.790961  2.061563  0.252736  1.422173 -1.139994  0.430935   
2172  2.337534  1.786388  1.998804  0.273980  0.994523 -1.137307  0.912597   
2173  2.305117  1.783949  2.031887  0.393376  0.282902 -1.128758  1.112527   
2174  2.305117  1.783238  1.988

In [2]:
[a,b]=df.shape
ntest=100
df_train = normalized_df.iloc[:(a-ntest), :]
df_test = normalized_df.iloc[(a-ntest):, :]
# train
Y = np.array(df_train['price'])
Y = np.expand_dims(Y, axis=1)
X0 = np.ones([a-ntest,1])
X1 = np.array(df_train[df_train.columns[df_train.columns!='price']])
X = np.concatenate([X0, X1],axis=1)
#test
Y_test = np.array(df_test['price'])
Y_test = np.expand_dims(Y_test, axis=1)
X0_test = np.ones([ntest,1])
X1_test= np.array(df_test[df_test.columns[df_test.columns!='price']])
X_test = np.concatenate([X0_test, X1_test],axis=1)

In [3]:
rho=1
from numpy.linalg import inv
time_start = time.clock()
beta_ls_exact= np.dot( np.dot(inv(np.dot(X.T,X) +rho*np.identity(b) ),X.T),Y)
time_elapsed = (time.clock() - time_start)
print('Time',time_elapsed)
print(beta_ls_exact)

('Time', 0.0055372776374282085)
[[-0.0115803 ]
 [ 0.36259694]
 [ 0.21770513]
 [-0.06965771]
 [ 0.36784859]
 [-0.14696702]
 [-0.00694018]
 [ 0.07745367]
 [-0.04210075]]


__b) (3 points) Estimate the value of the regression coefficients by using the function minimize
from the Python module Scipy.optimize. Try at least four available solvers and compare
their performance in terms of number of iterations, number of function, gradient and hessian
evaluations as well as total computational time.__


__SOLUTION__<br>
Now we're going to use different types of solvers (such as Nelder-Mead, Powell, dogleg and so on) in order to see which one best suits our problem.

In [4]:
from scipy.optimize import minimize
help(minimize)

Help on function minimize in module scipy.optimize._minimize:

minimize(fun, x0, args=(), method=None, jac=None, hess=None, hessp=None, bounds=None, constraints=(), tol=None, callback=None, options=None)
    Minimization of scalar function of one or more variables.
    
    In general, the optimization problems are of the form::
    
        minimize f(x) subject to
    
        g_i(x) >= 0,  i = 1,...,m
        h_j(x)  = 0,  j = 1,...,p
    
    where x is a vector of one or more variables.
    ``g_i(x)`` are the inequality constraints.
    ``h_j(x)`` are the equality constrains.
    
    Optionally, the lower and upper bounds for each element in x can also be
    specified using the `bounds` argument.
    
    Parameters
    ----------
    fun : callable
        Objective function.
    x0 : ndarray
        Initial guess.
    args : tuple, optional
        Extra arguments passed to the objective function and its
        derivatives (Jacobian, Hessian).
    method : str or callable, op

In [5]:
#ObjectiveFunction to minimize
def ridge_reg(beta_ls, X, Y,rho):
    beta_ls=np.matrix(beta_ls)
    z=Y-np.dot(X,beta_ls.T)
    return np.dot(z.T,z) + rho*np.dot(beta_ls,beta_ls.T)

And we will also need to define functions to obtain the gradient and the hessian, whose expressions are:<br>

__Gradient:__
$$ -2X' Y+2(X' X+\rho I)\beta $$
__Hessian:__
$$2(X'X+\rho\,I)$$

In [6]:
#Gradient:
def ridge_reg_der(beta_ls,X,Y,rho):
    beta_ls=np.matrix(beta_ls)
    pp=-2*np.dot((Y-np.dot(X,beta_ls.T)).T,X)+2*rho*beta_ls
    aa=np.squeeze(np.asarray(pp))
    return aa

#Hessian:
def ridge_reg_hess(beta_ls,X,Y,rho):
    return 2*np.dot(X.T,X)+2*rho*np.identity(b)


Let's start by __Nelder_Mead__ method: 

In [7]:
beta_ls0 = np.zeros(b)
time_start1 = time.clock()

res1 = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="Nelder-Mead",
                            options={"disp":True,"xtol":1e-10}) 
time_elapsed1 = (time.clock() - time_start1)

print('Time',time_elapsed1)
print (res1.x) 
print('Error',np.linalg.norm(beta_ls_exact.T-res1.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

('Time', 0.252950932074269)
[-0.04500509  0.31478912  0.12121656 -0.14988045  0.31996648 -0.24497396
 -0.07880544 -0.00367904  0.13425021]
('Error', 0.46041063550883876)


Here the warning message is indicating that the algorithm didn't converge.<br>
Let's see what happens by using the __Powell__ method:


In [8]:
beta_ls0 = np.zeros(b)
time_start2 = time.clock()

res2 = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="Powell",options={"disp":True,"xtol":1e-10}) 
time_elapsed2 = (time.clock() - time_start2)
print('Time',time_elapsed2)
print (res2.x) 
print('Error',np.linalg.norm(beta_ls_exact.T-res2.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.
         Current function value: 392.679092
         Iterations: 12
         Function evaluations: 2469
('Time', 0.462712122576161)
[[-0.01145566  0.36002798  0.22132933 -0.07012716  0.36645163 -0.14699988
  -0.00701157  0.07806219 -0.04311103]]
('Error', 0.0081789465690838472)


Let's check __Newton-CG__ as well:

In [9]:
beta_ls0 = np.zeros(b)
time_start3 = time.clock()

res3 = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="Newton-CG",
               jac=ridge_reg_der,hess=ridge_reg_hess,
               options={"disp":True,"xtol":1e-10}) 
time_elapsed3 = (time.clock() - time_start3)
print('Time',time_elapsed3)
print (res3.x) 
print('Error',np.linalg.norm(beta_ls_exact.T-res3.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.
         Current function value: 392.669371
         Iterations: 13
         Function evaluations: 14
         Gradient evaluations: 26
         Hessian evaluations: 13
('Time', 0.0065186105520594895)
[-0.0115803   0.36259694  0.21770513 -0.06965771  0.36784859 -0.14696702
 -0.00694018  0.07745367 -0.04210075]
('Error', 4.1130383837499984e-15)


Now let's try with other methods which comtemplates the hessian as well, such as __dogleg:__

In [10]:
beta_ls0 = np.zeros(b)
time_start4 = time.clock() 

res4 = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="dogleg",
               jac=ridge_reg_der,hess=ridge_reg_hess,
               options={"disp":True,"xtol":1e-10}) 
time_elapsed4 = (time.clock() - time_start4)
print('Time',time_elapsed4)
print (res4.x) 
print('Error',np.linalg.norm(beta_ls_exact.T-res4.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.
         Current function value: 392.669371
         Iterations: 1
         Function evaluations: 2
         Gradient evaluations: 2
         Hessian evaluations: 1
('Time', 0.0028394654551614495)
[-0.0115803   0.36259694  0.21770513 -0.06965771  0.36784859 -0.14696702
 -0.00694018  0.07745367 -0.04210075]
('Error', 8.2837542204893854e-15)


C:\Users\fra12\Anaconda2\lib\site-packages\scipy\optimize\_minimize.py:461: OptimizeWarning: Unknown solver options: xtol
  callback=callback, **options)


and __trust-ncg:__

In [11]:
beta_ls0 = np.zeros(b)
time_start5 = time.clock()

res5 = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="trust-ncg",
               jac=ridge_reg_der,hess=ridge_reg_hess,
               options={"disp":True,"xtol":1e-10}) 
time_elapsed5 = (time.clock() - time_start5)
print('Time',time_elapsed5)
print (res5.x) 
print('Error',np.linalg.norm(beta_ls_exact.T-res5.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.
         Current function value: 392.669371
         Iterations: 12
         Function evaluations: 13
         Gradient evaluations: 13
         Hessian evaluations: 12
('Time', 0.007332690204718828)
[-0.0115803   0.36259694  0.21770512 -0.06965772  0.3678486  -0.14696703
 -0.00694017  0.07745368 -0.04210075]
('Error', 1.6986604189212299e-08)


C:\Users\fra12\Anaconda2\lib\site-packages\scipy\optimize\_minimize.py:464: OptimizeWarning: Unknown solver options: xtol
  callback=callback, **options)


As the last step let's consider the __SLSQP__ which considers the Jacobian as well:

In [12]:
beta_ls0 = np.zeros(b)
time_start6 = time.clock() 
 
res6 = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="SLSQP",
               jac=ridge_reg_der,
               options={"disp":True,"xtol":1e-10}) 
time_elapsed6 = (time.clock() - time_start6)
print('Time',time_elapsed6)
print (res6.x) 
print('Error',np.linalg.norm(beta_ls_exact.T-res6.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.    (Exit mode 0)
            Current function value: 392.669371103
            Iterations: 12
            Function evaluations: 45
            Gradient evaluations: 12
('Time', 0.04534270065378121)
[-0.0115809   0.36259591  0.2177053  -0.0696582   0.36784794 -0.14696801
 -0.00694075  0.07745273 -0.0420991 ]
('Error', 4.488789372649654e-06)


C:\Users\fra12\Anaconda2\lib\site-packages\ipykernel_launcher.py:6: OptimizeWarning: Unknown solver options: xtol
  


Finally let's compare the results:

In [13]:
method =["Nelder_Mead","Powell","Newton-CG","dogleg","trust-ncg","SLSQP"]
it= [res1.nit, res2.nit, res3.nit, res4.nit, res5.nit, res6.nit]
fun =[res1.fun, res2.fun, res3.fun, res4.fun, res5.fun, res6.fun]
time=[time_elapsed1, time_elapsed2, time_elapsed3, time_elapsed4, time_elapsed5, time_elapsed6]
hes=["","",res3.nhev,"",res5.nhev,""]
jac=["", "", "", "", "", res6.njev]


In [14]:
print('Methods:', method)
print('Iterations:',it)
print('ObjectiveFunction values:', fun)
print('Time:', time)
print('Hessian:', hes)
print('Jacobian', jac)

('Methods:', ['Nelder_Mead', 'Powell', 'Newton-CG', 'dogleg', 'trust-ncg', 'SLSQP'])
('Iterations:', [1236, 12, 13, 1, 12, 12])
('ObjectiveFunction values:', [406.08780298319897, matrix([[ 392.67909216]]), matrix([[ 392.6693711]]), matrix([[ 392.6693711]]), matrix([[ 392.6693711]]), 392.6693711025743])
('Time:', [0.252950932074269, 0.462712122576161, 0.0065186105520594895, 0.0028394654551614495, 0.007332690204718828, 0.04534270065378121])
('Hessian:', ['', '', 13, '', 12, ''])
('Jacobian', ['', '', '', '', '', 12])


We can see that all the methods converge by giving a solution to our optimization problem and that the one that performs better in terms of iterations is dogleg with only 1 of them.<br> 
Then, apart from the Nelder_Mead method whose iterations are much higher that the rest (1236), among the other methods the quickest are Newton-CG (with 13 iterations) and SLSQP (with 12 iterations).

__c) Estimate the value of the regression coefficients by implementing the:<br>
i. (1.5 points) Gradient method.<br>
ii. (1.5 points) Newton method.<br>
Consider a line search technique to improve the algorithm convergence, e.x., Armijo rule. <br>
Compare the performance of these algorithms (number of iterations and total computational
time).__

__SOLUTION__<br>
__i.) Gradient Method (with Wolfe Conditions)__ <br>

$\bullet$ Initial iteratation $x_0$

$\bullet$ Computes search (descent) directions $p_k=-\nabla f(x_k)$

$\bullet$ Far from the solution, computes a steplength $\alpha_k>0$, to which Armijo rule is applied:

$$ f(x_k+\alpha_k p_k)\leq f(x_k)+c_1 \alpha_k\nabla f(x_k)' p_k$$   

$\bullet$ Movement:
$$x_{k+1} = x_k + \alpha_k\ p_k$$
Until convergence to a local solution.

In [15]:
import time
(a,b)=X.shape
beta_lsg=np.zeros(b)
n_iter=5000
OF_iter=np.zeros(n_iter)
tol_iter=np.zeros(n_iter)
alpha_iter=np.zeros(n_iter)
i=0;
tol=10000;
epsilon=1e-3;

time_start7 = time.clock()
while (i <= n_iter-2) and (tol>epsilon):
    i=i+1
    grad= ridge_reg_der(beta_lsg,X,Y,rho)
    ddirect=-grad
    # Armijo condition
    #######################################
    sigma=0.1
    beta=0.1
    alpha=1
    while (ridge_reg(beta_lsg+alpha*ddirect,X,Y,rho)>
           ridge_reg(beta_lsg,X,Y,rho)+alpha*sigma*np.dot(ddirect,grad)) :
        alpha=alpha*beta
    #######################################
    beta_lsg= beta_lsg +alpha*ddirect
    OF_iter[i]=ridge_reg(beta_lsg, X, Y,rho)
    tol=np.linalg.norm(grad,ord=2)
    tol_iter[i]=tol
    alpha_iter[i]=alpha

time_elapsed7 = (time.clock() - time_start7) 
print('Time:',time_elapsed7)
print('Iterations:',i)
print('ObjectiveFunction value',OF_iter[i])
print("Computed Betas:",beta_lsg)
print("Exact Betas:",np.transpose(beta_ls_exact))
print('Tolerance:',tol)
print('Error',np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2))

('Time:', 0.18286200197887914)
('Iterations:', 199)
('ObjectiveFunction value', 392.66937110236034)
('Computed Betas:', array([-0.0115803 ,  0.36259619,  0.21770316, -0.06965934,  0.36784741,
       -0.14696808, -0.00694137,  0.0774533 , -0.04209708]))
('Exact Betas:', array([[-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
        -0.14696702, -0.00694018,  0.07745367, -0.04210075]]))
('Tolerance:', 0.00087598351475383915)
('Error', 8.3996841082041457e-06)


__i.) Newton Method (with Wolfe Conditions)__ <br>

$\bullet$ Initial iteratation $x_0$

$\bullet$ Computes search (descent) directions  $p_k=-(\nabla^2 f(x_k))^{-1} \nabla f(x_k)$ where $\nabla^2 f(x_k)$ is
nonsingular. 

$\bullet$ Far from the solution, computes a steplength $\alpha_k>0$, to which Armijo rule is applied:

$$ f(x_k+\alpha_k p_k)\leq f(x_k)+c_1 \alpha_k\nabla f(x_k)' p_k$$   

$\bullet$ Movement:
$$x_{k+1} = x_k + \alpha_k\ p_k$$
Until convergence to a local solution.

In [16]:
import time
(a,b)=X.shape
beta_lsn=np.zeros(b) 
n_iter=5000 
OF_iter2=np.zeros(n_iter)
tol_iter=np.zeros(n_iter)
alpha_iter=np.zeros(n_iter)
j=0;
tol=10000;
epsilon=1e-3;

time_start8 = time.clock()
while (j <= n_iter-2) and (tol>epsilon):
    j=j+1
    grad= ridge_reg_der(beta_lsg,X,Y,rho)
    hess= ridge_reg_hess(beta_lsn,X,Y,rho)
    ddirect= -np.dot(inv(hess),grad)
     # Armijo condition (opcionales - tamaño de paso variable)
    #######################################
    sigma=0.1
    beta=0.1
    alpha=1
    while (ridge_reg(beta_lsg+alpha*ddirect,X,Y,rho)>
           ridge_reg(beta_lsg,X,Y,rho)+alpha*sigma*np.dot(ddirect,grad)) :
        alpha=alpha*beta
    beta_lsg= beta_lsg +alpha*ddirect
    OF_iter2[j]=ridge_reg(beta_lsg, X, Y,rho)
    tol=np.absolute((OF_iter2[j]-OF_iter2[j-1])/OF_iter2[j]) #tolerance
    tol=np.linalg.norm(grad,ord=2)
    tol_iter[j]=tol
    alpha_iter[j]=alpha

time_elapsed8 = (time.clock() - time_start8) 
print('Time',time_elapsed8)
print('Iterations',j)
print('ObjectiveFunction value',OF_iter2[j])
print("Computed Betas:",beta_lsg)
print("Exact Betas",np.transpose(beta_ls_exact))
print('Tolerance=',tol)
print('Error',np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2))


('Time', 0.001966505827624143)
('Iterations', 2)
('ObjectiveFunction value', 392.66937110033604)
('Computed Betas:', array([-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
       -0.14696702, -0.00694018,  0.07745367, -0.04210075]))
('Exact Betas', array([[-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
        -0.14696702, -0.00694018,  0.07745367, -0.04210075]]))
('Tolerance=', 2.4869395540680865e-13)
('Error', 3.7313564658674407e-15)


In [17]:
time=[time_elapsed7,time_elapsed8]
fun=[OF_iter[i],OF_iter2[j]]
it=[i,j]
print('Objective Function Value:',fun)
print('Iterations:',it)
print('Time:',time)


('Objective Function Value:', [392.66937110236034, 392.66937110033604])
('Iterations:', [199, 2])
('Time:', [0.18286200197887914, 0.001966505827624143])


We can see that Netwton's methods performs faster than the Gradient descending method due to the number of iterations which is 2 compared to the 199 of the Gradient method.

__d) Estimate the value of the regression coefficients by implementing the:<br>
i. Coordinate gradient method.<br>
ii. Stochastic gradient method.__<br>

__SOLUTION:<br>
i. Coordinate gradient method.__<br>

$\bullet$  From an initial iterate $x_0$

$\bullet$ Compute search (descent) directions. Find $i_k \in {1,...,p}$ such that $\nabla_{i_k} f(x_k)$ is maximum. $p_k=- \nabla_{i_k} f(x_k) e_{i_k}$

$\bullet$ Far from the solution, compute a steplength $\alpha_k>0$:

$\bullet$ Movement:
$$x_{k+1} = x_k + \alpha_k\ p_k$$
Until convergence to a local solution.


In [18]:
import time
(a,b)=X.shape
beta_lsg=np.zeros(b) 

alpha=0.0001
n_iter=5000 
OF_iter3=np.zeros(n_iter)
tol_iter=np.zeros(n_iter)
alpha_iter=np.zeros(n_iter)
k=0;
tol=10000;
epsilon=1e-3;

time_start9 = time.clock()
while (k <= n_iter-2) and (tol>epsilon):
    k=k+1
    grad= ridge_reg_der(beta_lsg,X,Y,rho)
    ddirect=-np.eye(1,len(grad),np.argmax(abs(grad)))*grad
    
    beta_lsg= beta_lsg +alpha*ddirect
    OF_iter3[k]=ridge_reg(beta_lsg, X, Y,rho)
    tol=np.linalg.norm(grad,ord=2)
    tol_iter[k]=tol
    alpha_iter[k]=alpha
    

time_elapsed9 = (time.clock() - time_start9) 
print('Time',time_elapsed9)
print('Iterations',k)
print('ObjectiveFunction value',OF_iter3[k])
print("Computed Betas:",beta_lsg)
print("Exact Betas",np.transpose(beta_ls_exact))
print('Tolerance=',tol)
print('Error',np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2))

('Time', 0.261903248254614)
('Iterations', 2050)
('ObjectiveFunction value', 392.66937110284016)
('Computed Betas:', array([[-0.0115802 ,  0.36259574,  0.2177029 , -0.06965961,  0.36784715,
        -0.14696837, -0.00694162,  0.07745311, -0.04209661]]))
('Exact Betas', array([[-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
        -0.14696702, -0.00694018,  0.07745367, -0.04210075]]))
('Tolerance=', 0.00099418722580234568)
('Error', 9.7917516111327325e-06)


__ii. Stochastic gradient method.__<br>

$\bullet$ From an initial iterate $x_0$

$\bullet$ Compute search (descent) directions. Choose $j_k \in {1,...,n}$ at random $p_k=- \nabla f_{j_k}(x_k) $

$\bullet$ Far from the solution, compute a steplength $\alpha_k>0$:

$\bullet$ Movement:
$$x_{k+1} = x_k + \alpha_k\ p_k$$
Until convergence to a local solution.

In [19]:
import random

#Random Gradient:
def ridge_reg_der2(beta_ls,X,Y,rho):
    j=int(random.uniform(1,len(Y))) 
    beta_ls=np.matrix(beta_ls)
    pp=-2*(Y[j]-np.dot(X[j,:],beta_ls.T)).T*X[j,:]+2*rho*beta_ls
    aa=np.squeeze(np.asarray(pp))
    return aa

In [20]:
(a,b)=X.shape
beta_lsg=np.zeros(b) 

alpha=0.0001
n_iter=5000 
OF_iter4=np.zeros(n_iter)
tol_iter=np.zeros(n_iter)
alpha_iter=np.zeros(n_iter)
z=0;
tol=10000;
epsilon=1e-3;

time_start10 = time.clock()
while (z <= n_iter-2) and (tol>epsilon):
    z=z+1
    grad= ridge_reg_der2(beta_lsg,X,Y,rho)
    ddirect=-grad
    beta_lsg= beta_lsg +alpha*ddirect
    OF_iter4[z]=ridge_reg(beta_lsg, X, Y,rho)
    tol=np.linalg.norm(grad,ord=2)
    tol_iter[z]=tol
    alpha_iter[z]=alpha
    

time_elapsed10 = (time.clock() - time_start10) 
print('Time',time_elapsed10)
print('Iterations',z)
print('ObjectiveFunction value',OF_iter4[z])
print("Computed Betas:",beta_lsg)
print("Exact Betas",np.transpose(beta_ls_exact))
print('Tolerance=',tol)
print('Error',np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2))

('Time', 0.4990551470698039)
('Iterations', 4999)
('ObjectiveFunction value', 530.50052983961052)
('Computed Betas:', array([-0.02265276,  0.18123517,  0.16140898, -0.09454586,  0.14721616,
       -0.10150714, -0.01648385,  0.10464204,  0.11578218]))
('Exact Betas', array([[-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
        -0.14696702, -0.00694018,  0.07745367, -0.04210075]]))
('Tolerance=', 4.1560268662383706)
('Error', 0.9522483562836368)


In [21]:
time=[time_elapsed9,time_elapsed10]
fun=[OF_iter3[k],OF_iter4[z]]
it=[k,z]
print('Objective Function Value:',fun)
print('Iterations:',it)
print('Time:',time)


('Objective Function Value:', [392.66937110284016, 530.50052983961052])
('Iterations:', [2050, 4999])
('Time:', [0.261903248254614, 0.4990551470698039])


The results given by Coordinate Gradient method are better in terms of fewer iterations (2050 vs 4999 of the Stochastic method) and time elapsed. Moreover the error given by the Stochastic method is much higher and the algorithm actually converges to a different minimum compared to all the previous methods.